controllers/provisioning_controller.go

```diff
@@ -153,39 +153,45 @@ func (r *ProvisioningReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error
 		return ctrl.Result{}, nil
 	}
 
-	// examine DeletionTimestamp to determine if object is under deletion
-	if baremetalConfig.ObjectMeta.DeletionTimestamp.IsZero() {
-		// The object is not being deleted, so if it does not have our finalizer,
-		// then lets add the finalizer and update the object. This is equivalent
-		// registering our finalizer.
-		if !containsString(baremetalConfig.ObjectMeta.Finalizers,
-			metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add finalizer becasue it doesn't already exist
-			baremetalConfig.ObjectMeta.Finalizers = append(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to update Provisioning CR with its finalizer")
-			}
-		}
-	} else {
-		// The Provisioning object is being deleted
-		if containsString(baremetalConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
-			// Add any specific deletion logic here
-
-			// Remove our finalizer from the list and update it.
-			baremetalConfig.ObjectMeta.Finalizers = removeString(baremetalConfig.ObjectMeta.Finalizers,
-				metal3iov1alpha1.ProvisioningFinalizer)
-			if err := r.Client.Update(context.Background(), baremetalConfig); err != nil {
-				return ctrl.Result{}, errors.Wrap(err, "failed to remove finalizer from Provisioning CR")
-			}
+	// Read container images from Config Map
+	var containerImages provisioning.Images
+	if err := provisioning.GetContainerImages(&containerImages, r.ImagesFilename); err != nil {
+		// Images config map is not valid
+		// Provisioning configuration is not valid.
+		// Requeue request.
+		r.Log.Error(err, "invalid contents in images Config Map")
+		co_err := r.updateCOStatus(ReasonInvalidConfiguration, err.Error(), "invalid contents in images Config Map")
```

Comment:
all these calls to updateCOStatus() can fail if the CO does not exist. Can you please move the ensureClusterOperator() to above the first possible call to updateCOStatus?

Reply:
Updated. Please let me know if you are fine with the current amount of refactoring done within Reconcile() so that it doesn't get bloated.